src/engine/plugins/isochrone.cpp

```diff
@@ -0,0 +1,330 @@
+                                  reverse_duration_offset,
+                                  ucoord,
+                                  ucoord);
+        }
+    });
+
+    std::vector<std::size_t> sources;
+    std::vector<std::size_t> destinations;
+
+    sources.push_back(0);
+    destinations.resize(phantoms.size() - 1);
+    std::iota(destinations.begin(), destinations.end(), 1);
+
+    TIMER_STOP(PHANTOM_TIMER);
+    util::Log() << "Make phantom nodes " << TIMER_MSEC(PHANTOM_TIMER);
+
+    // Phase 1 - outgoing, upwards dijkstra search, setting d(v) for all v we visit
+    heaps.InitializeOrClearFirstThreadLocalStorage(facade->GetNumberOfNodes());
+    heaps.InitializeOrClearSecondThreadLocalStorage(facade->GetNumberOfNodes());
+
+    auto &query_heap = *(heaps.forward_heap_1);
+    if (phantoms[0].forward_segment_id.enabled)
+    {
+        query_heap.Insert(phantoms[0].forward_segment_id.id,
+                          -phantoms[0].GetForwardWeightPlusOffset(),
+                          phantoms[0].forward_segment_id.id);
+    }
+    if (phantoms[0].reverse_segment_id.enabled)
+    {
+        query_heap.Insert(phantoms[0].reverse_segment_id.id,
+                          -phantoms[0].GetReverseWeightPlusOffset(),
+                          phantoms[0].reverse_segment_id.id);
+    }
+
+    while (query_heap.Size() > 0)
+    {
+        const NodeID node = query_heap.DeleteMin();
+        const EdgeWeight weight = query_heap.GetKey(node);
+
+        for (auto edge : facade->GetAdjacentEdgeRange(node))
+        {
+            const auto &data = facade->GetEdgeData(edge);
+            if (data.forward)
+            {
+                const NodeID to = facade->GetTarget(edge);
+                const EdgeWeight edge_weight = data.weight;
+
+                BOOST_ASSERT_MSG(edge_weight > 0, "edge_weight invalid");
+                const EdgeWeight to_weight = weight + edge_weight;
+
+                // No need to search further than our maximum radius
+                if (weight > range * 10)
+                    continue;
+
+                // New Node discovered -> Add to Heap + Node Info Storage
+                if (!query_heap.WasInserted(to))
+                {
+                    query_heap.Insert(to, to_weight, node);
+                }
+                // Found a shorter Path -> Update weight
+                else if (to_weight < query_heap.GetKey(to))
+                {
+                    // new parent
+                    query_heap.GetData(to) = node;
+                    query_heap.DecreaseKey(to, to_weight);
+                }
+            }
+        }
+    }
+
+    // Phase 2 - scan nodes in descending CH order, updating d(v) where we can
+    TIMER_START(TABLE_TIMER);
+    std::sort(phantoms.begin(), phantoms.end(), [](const auto &a, const auto &b) {
+        const auto &a_id =
+            a.forward_segment_id.enabled ? a.forward_segment_id.id : a.reverse_segment_id.id;
```

Comment:
Sadly the node ID is not a proxy for contraction level in our CH implementation: We keep the node ID constant starting in the `EdgeBasedGraphFactory`, because we need to save the ID in the RTree which happens before contraction.

The way you could fix this would be to load the `.osrm.level` file which contains exactly that (we use it for speeding up the contraction on successive contractions).